aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/ObserveQueryExecutor.java

```diff
@@ -132,42 +132,50 @@ public void observeQuery(
 
-            @SuppressWarnings("unchecked") 
+            @SuppressWarnings("unchecked")
             StorageItemChange<T> itemChanged = (StorageItemChange<T>) value;
             try {
                 if (sqlQueryProcessor.modelExists(itemChanged.item(), options.getQueryPredicate())) {
                     updateCompleteItemMap(itemChanged);
                 } else if (itemChanged.type() == StorageItemChange.Type.UPDATE) {
                     completeItemMap.remove(itemChanged.item().getPrimaryKeyString());
+                } else if (itemChanged.type() == StorageItemChange.Type.DELETE) {
+                    completeItemMap.remove(itemChanged.item().getPrimaryKeyString());
                 }
                 collect(itemChanged, onQuerySnapshot, itemClass, options, onObservationError);
             } catch (DataStoreException exception) {
                 onObservationError.accept(exception);
             }
         };
-        threadPool.submit(() -> queryLocalData(itemClass, options, onQuerySnapshot, onObservationError));
-
-        disposable = itemChangeSubject
-            .filter(x -> x.item().getClass().isAssignableFrom(itemClass))
-            .subscribe(
-                onItemChanged::accept,
-                failure -> {
-                    if (failure instanceof DataStoreException) {
-                        onObservationError.accept((DataStoreException) failure);
-                        return;
-                    }
-                    onObservationError.accept(new DataStoreException(
-                            "Failed to observe items in storage adapter.",
-                            failure,
-                            "Inspect the failure details."
-                    ));
-                },
-                onObservationComplete::call
-            );
+        threadPool.submit(() -> queryLocalData(
+            itemClass,
+            options,
+            value -> {
+                disposable = itemChangeSubject
+                        .filter(x -> x.item().getClass().isAssignableFrom(itemClass))
+                        .subscribe(
+                                onItemChanged::accept,
+                                failure -> {
+                                    if (failure instanceof DataStoreException) {
+                                        onObservationError.accept((DataStoreException) failure);
+                                        return;
+                                    }
+                                    onObservationError.accept(new DataStoreException(
+                                            "Failed to observe items in storage adapter.",
+                                            failure,
+                                            "Inspect the failure details."
+                                    ));
+                                },
+                                onObservationComplete::call
+                        );
+                onObservationStarted.accept(this);
```

Comment:
🤔 Is it strictly necessary for the `onObservationStarted` to be so late?

There is a slight behaviour change here in that if the initial query fails the caller will never see the onStarted callback, only onError (prior behaviour was to see onStarted and then onError). I'm not sure that is a problem, in fact it may be more correct than the prior implementation.

I was also looking at whether it would be possible for `onItemChange` to get invoked prior on `onObservationStarted`, it's a bit ambiguous to me but I don't think so?

Reply:
I had requested this change after diving in together because it was easy to replicate incorrect values and counts if mutations were made before the initial query was successful. Once we report `onObservationStarted`, my expectation is that no observations would be missed once that value is returned. This was not the case with the immediate callback.

If `completeItemMap` is empty (which is the case before its populated with initial values from a successful query, any item changes (such as a delete) will not properly be updated/removed from competeItemMap.

Reply:
Makes sense, thanks.